source/common/lua/wrappers.h

```diff
@@ -35,5 +35,64 @@ class BufferWrapper : public BaseLuaObject<BufferWrapper> {
   const Buffer::Instance& data_;
 };
 
+class MetadataMapWrapper;
+
+/**
+ * Iterator over a metadata map.
+ */
+class MetadataMapIterator : public BaseLuaObject<MetadataMapIterator> {
+public:
+  MetadataMapIterator(MetadataMapWrapper& parent);
+
+  static ExportedFunctions exportedFunctions() { return {}; }
+
+  DECLARE_LUA_CLOSURE(MetadataMapIterator, luaPairsIterator);
+
+private:
+  MetadataMapWrapper& parent_;
+  ProtobufWkt::Map<std::string, ProtobufWkt::Value>::const_iterator current_;
+};
+
+/**
+ * Lua wrapper for a metadata map.
+ */
+class MetadataMapWrapper : public Envoy::Lua::BaseLuaObject<MetadataMapWrapper> {
+public:
+  MetadataMapWrapper(const ProtobufWkt::Struct& metadata) : metadata_{metadata} {}
+
+  static ExportedFunctions exportedFunctions() {
+    return {{"get", static_luaGet}, {"__pairs", static_luaPairs}};
+  }
+
+private:
+  /**
+   * Get a metadata value from the map.
+   * @param 1 (string): filter.
+   * @return string value if found or nil.
+   */
+  DECLARE_LUA_FUNCTION(MetadataMapWrapper, luaGet);
+
+  /**
+   * Implementation of the __pairs metamethod so a metadata wrapper can be iterated over using
+   * pairs().
+   */
+  DECLARE_LUA_FUNCTION(MetadataMapWrapper, luaPairs);
+
+  // Envoy::Lua::BaseLuaObject
+  void onMarkDead() override {
+    // Iterators do not survive yields.
+    iterator_.reset();
+  }
+
+  void setValue(lua_State* state, const ProtobufWkt::Value&& value);
```

Comment:
Unless I'm missing something I don't think passing R-value (`&&`) here and to `createTable` is useful (and makes code a bit more confusing). I think you just want to pass a normal const reference?